service/src/main/java/bio/terra/workspace/service/resource/controlled/cloud/gcp/dataproccluster/CreateDataprocClusterStep.java

```diff
@@ -0,0 +1,345 @@
+import com.google.api.services.dataproc.model.AcceleratorConfig;
+import com.google.api.services.dataproc.model.AutoscalingConfig;
+import com.google.api.services.dataproc.model.Cluster;
+import com.google.api.services.dataproc.model.ClusterConfig;
+import com.google.api.services.dataproc.model.DiskConfig;
+import com.google.api.services.dataproc.model.EndpointConfig;
+import com.google.api.services.dataproc.model.GceClusterConfig;
+import com.google.api.services.dataproc.model.InstanceGroupConfig;
+import com.google.api.services.dataproc.model.LifecycleConfig;
+import com.google.api.services.dataproc.model.Operation;
+import com.google.api.services.dataproc.model.SoftwareConfig;
+import com.google.common.annotations.VisibleForTesting;
+import java.io.IOException;
+import java.time.Duration;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import org.apache.commons.lang3.StringUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.http.HttpStatus;
+
+/**
+ * A step for creating a Dataproc cluster in the Google cloud.
+ *
+ * <p>Undo deletes the created cluster.
+ */
+public class CreateDataprocClusterStep implements Step {
+  private final Logger logger = LoggerFactory.getLogger(CreateDataprocClusterStep.class);
+  private final ControlledResourceService controlledResourceService;
+  private final ControlledDataprocClusterResource resource;
+  private final String petEmail;
+  private final String workspaceUserFacingId;
+  private final CrlService crlService;
+  private final CliConfiguration cliConfiguration;
+
+  public CreateDataprocClusterStep(
+      ControlledResourceService controlledResourceService,
+      ControlledDataprocClusterResource resource,
+      String petEmail,
+      String workspaceUserFacingId,
+      CrlService crlService,
+      CliConfiguration cliConfiguration) {
+    this.petEmail = petEmail;
+    this.resource = resource;
+    this.workspaceUserFacingId = workspaceUserFacingId;
+    this.controlledResourceService = controlledResourceService;
+    this.crlService = crlService;
+    this.cliConfiguration = cliConfiguration;
+  }
+
+  @Override
+  public StepResult doStep(FlightContext flightContext)
+      throws InterruptedException, RetryException {
+    final GcpCloudContext gcpCloudContext =
+        flightContext
+            .getWorkingMap()
+            .get(ControlledResourceKeys.GCP_CLOUD_CONTEXT, GcpCloudContext.class);
+    String projectId = gcpCloudContext.getGcpProjectId();
+
+    String region =
+        FlightUtils.getRequired(
+            flightContext.getWorkingMap(), CREATE_RESOURCE_REGION, String.class);
+
+    ApiGcpDataprocClusterCreationParameters creationParameters =
+        flightContext
+            .getInputParameters()
+            .get(CREATE_DATAPROC_CLUSTER_PARAMETERS, ApiGcpDataprocClusterCreationParameters.class);
+
+    ControlledGcsBucketResource stagingBucket =
```

Comment:
Maybe use `FlightUtils.getRequired` for the buckets, since they cannot be missing.